### ISSUE TITLE
GPU usage decreasing with epochs

### ISSUE THREAD

**Issue by YonyBresler** [PyTorchLightning/pytorch-lightning#581]
I'm having a similar issue to #373 with gradual slowing down while training.

The model works fine but gradually with each epoch GPU utilization is decreasing, and the time needed to complete it increases.
<img width="652" alt="Screen Shot 2019-12-03 at 6 23 33 PM" src="https://user-images.githubusercontent.com/24940683/70098507-1d8cde80-15fa-11ea-88ce-f8f473bfd540.png">
GPU memory usage is also slowly decreasing at the same rate, so I don't think there's a memory leak, and i/o usage is roughly constant at 1200 K/s, well below the sustained speed my HD will support. CPU usage is 20-40%, total CPU memory is <30%, and GPU usage, power and temperature are all in regime where there should be no throttling, so I don't think it's a hardware issue.

Stopping the run, letting things cool down, and restarting with loaded weights doesn't revert speed, and continues with the same slower speed. 

I don't have these issues if I replace lightning with a basic training loop.

Has there been any other indication as to what might be causing this?

**Comment by YonyBresler** [comment]
On closer look it seems I/O does increase as epochs roll on, trying to figure out what is writing to disk so much since aside from default logging my model doesn't write anything, and epochs are long so weights shouldn't be an issue either.

**Comment by williamFalcon** [comment]
maybe the tensorboard logger is doing that? it’s the only thing that would write

**Comment by YonyBresler** [comment]
Thanks, the writes aren't to the local folder during an epoch, but it's running within an nvidia-docker, so there's some filesystem writes but it's a bit obscured where exactly.

**Comment by YonyBresler** [comment]
The I/O issue seemed to be an issue between tqdm and screen, running the model outside of screen eliminates the high IO load, back to normal < 1M/s.

The gradual slowing with epochs remains, so I think there's another issue, I'll try to test more to narrow it down, but if you have any suggestions, I'm happy to follow up if I can.

**Comment by williamFalcon** [comment]
got it. you can turn off thr prog bar in the trainer as an arg

**Comment by YonyBresler** [comment]
Avoiding screen seems to solve the I/O issue, but to be safe I restarted with turned off progress bar, there is now almost zero writing to disk, yet GPU usage drops to 30% during epochs (starts at 70%!), though interestingly during validation step usage jumps back up to 60%, so it must be something related to back-propogation (since I evaluate the same loss during validation).

Any other ideas for what I can test? I'm testing lightning for us to use it, other than this issue everything else works great, would like to resolve it so we can migrate.

**Comment by sneiman** [comment]
I have also experienced the problem of training slowing per epoch. I found a few causes to be aware of ...

Pytorch tensors will accumulate history so you have to be careful to NOT create a tensor outside of training/val/test loops that accumulate history, or to append to evergrowing lists unnecessarily. See this for discussion: https://discuss.pytorch.org/t/training-gets-slow-down-by-each-batch-slowly/4460
This was not my problem, unfortunately.

In the above discussion, the recommendation is also made to periodically make this call:
torch.cuda.empty_cache()
I put this in on_epoch_start() - and almost all of my problems disappeared.

However, I still have an important amount of this when using torch.optim.lr_scheduler.ReduceLROnPlateau(). Each time the learning rate is reduced the epoch time goes up by about 2 seconds in my case. The increase starts following the first time lr is reduced, persists, and increases again each time the lr is reduced. 

I have not tried any other lr schedulers, and am just starting to track down if there is something in ReduceLROnPlateau() or lightning. If I learn any more, I will post here.

**Comment by williamFalcon** [comment]
empty cache will make your code really slow... we make sure to release all tensors, so look where you store any tensors to make sure you do too

**Comment by sneiman** [comment]
Got it. Have removed - def a few seconds quicker per epoch. And do not see the slowdown EXCEPT with lr scheduler as above. I looked at the scheduler code and nothing jumps out. All tensors are properly managed. thx

**Comment by YonyBresler** [comment]
Thanks for the info, I'll also try to see if I have any tensors stored outside that loop that are getting appended to and see if that resolves this issue

**Comment by sneiman** [comment]
The pytorch optimizer documentation says that optimizers should be constructed **after** moving the model to gpu. Inspection of dp_mixin.py shows them being constructed before, however. Could this be the problem?

**Comment by YonyBresler** [comment]
I have also done a similar exercise and the code does not slow down with epochs when using it with my own training loop. Hope this can be further investigated and resolved

**Comment by huntzhan** [comment]
hi @williamFalcon , I'm experiencing a similar GPU utilization issue, dropping from 20% to 10% in ~20k steps. Is there any recommended way to set up performance profiling in pytorch-lighting?

**Comment by huntzhan** [comment]
After setting `logger=False`, the issue has been mitigated. Previously the `save_dir` of `TestTubeLogger` was set to a path in a shared folder, hence the high-frequency logging file I/O becomes the bottleneck.

**Comment by sneiman** [comment]
setting logger to False did not work for me. still losing ~1-1.5 secs per epoch. does NOT appear to be related to lr scheduler at this point.

**Comment by YonyBresler** [comment]
My issues are also not related to scheduler, logger or hardware.

My best guess: Something in the model is saving gradients that aren't needed, though I can't figure out what. Trying another (simpler) model on the same machine and libraries shows no slowdown.

**Comment by williamFalcon** [comment]
@YonyBresler any updates?

**Comment by YonyBresler** [comment]
@williamFalcon haven't been able to narrow down the issue, beyond the following:

It's not logging or hardware related, but with a complex model, something is causing a slowdown per epoch during training, likely some graph history being accumulated, as suggested above.

Resuming from a saved run will 'remember' the slowdown, changing from lightning to a simple custom training loop eliminates the issue, and using a simpler model with lightning doesn't have this issue on the same hardware & software stack.

Unfortunately I can't paste my complex model here, but I'll try to look through a debugger and see if I can dig up anything else, but it sounds like @sneiman is running into the exact same issue as I am, so I hope it should be reproducible.

**Comment by sneiman** [comment]
FWIW - this is not model related, as I have run the exact same model with my own training loop without seeing any loss of performance. I have been doing some work comparing dp and ddp, and my current effort shows the loss of performance ONLY with dp.

**Comment by williamFalcon** [comment]
how's this going?
dp is very slow btw compared with ddp

**Comment by sneiman** [comment]
I haven't seen is issue in .60 or master.

Sent from mobile device - sorry for typos and brevity.

This entire message is confidential. If it isn't intended for you, you may not use it - so please throw it away and forget about it.

On Mar 6, 2020, at 4:28 PM, William Falcon <notifications@github.com> wrote:

﻿

how's this going?
dp is very slow btw compared with ddp

**Comment by Borda** [comment]
Any update here? feel free t reopen if needed... :robot:

**Comment by congvm-it** [comment]
@sneiman  @williamFalcon  Did you find any solutions for this issue? I tried all those things you mentioned above but GPU utilization gradually drop down every iterations.

**Comment by kklemon** [comment]
I'm having the same problem, which slowly drives me crazy as I can't find the cause.

The complexity of my model and training setup makes investigating the problem not easy, but here a list what I tried to so far without success:

- Replacing the actual data module with a toy one to exclude I/O issues
- Setting `num_workers` in the dataloader to 0 (no subprocesses)
- Deactivating model checkpointing
- Deactivating the W&B logger
- Completely resetting the env and updating all packages

@YonyBresler @sneiman @huntzhan @congvmit, were you able to resolve the issue in the end and if so, what was the cause? Could you maybe provide the different components (loggers, callbacks, lr schedulers, etc.) that you used within your training setup?
